shell/browser/renderer_host/electron_render_message_filter.cc

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#include "shell/browser/renderer_host/electron_render_message_filter.h"
+
+#include <stdint.h>
+
+#include <memory>
+#include <string>
+
+#include "base/bind.h"
+#include "base/bind_helpers.h"
+#include "base/logging.h"
+#include "base/stl_util.h"
+#include "base/task/post_task.h"
+#include "chrome/browser/browser_process.h"
+#include "chrome/browser/chrome_notification_types.h"
+#include "chrome/browser/predictors/preconnect_manager.h"
+#include "components/network_hints/common/network_hints_common.h"
+#include "components/network_hints/common/network_hints_messages.h"
+#include "content/public/browser/browser_task_traits.h"
+#include "content/public/browser/notification_service.h"
+#include "content/public/browser/render_process_host.h"
+#include "content/public/browser/storage_partition.h"
+#include "shell/browser/api/atom_api_session.h"
+#include "shell/browser/net/preconnect_manager_factory.h"
+
+using content::BrowserThread;
+
+namespace {
+
+const uint32_t kRenderFilteredMessageClasses[] = {
+    NetworkHintsMsgStart,
+};
+
+}  // namespace
+
+ElectronRenderMessageFilter::ElectronRenderMessageFilter(
+    electron::api::Session* session)
+    : BrowserMessageFilter(kRenderFilteredMessageClasses,
+                           base::size(kRenderFilteredMessageClasses)),
+      session_(session) {}
+
+ElectronRenderMessageFilter::~ElectronRenderMessageFilter() {}
+
+bool ElectronRenderMessageFilter::OnMessageReceived(
+    const IPC::Message& message) {
+  bool handled = true;
+  IPC_BEGIN_MESSAGE_MAP(ElectronRenderMessageFilter, message)
+    IPC_MESSAGE_HANDLER(NetworkHintsMsg_Preconnect, OnPreconnect)
+    IPC_MESSAGE_UNHANDLED(handled = false)
+  IPC_END_MESSAGE_MAP()
+
+  return handled;
+}
+
+void ElectronRenderMessageFilter::OnPreconnect(const GURL& url,
+                                               bool allow_credentials,
+                                               int count) {
+  if (count < 1) {
+    LOG(WARNING) << "NetworkHintsMsg_Preconnect IPC with invalid count: "
+                 << count;
+    return;
+  }
+
+  if (!url.is_valid() || !url.has_host() || !url.has_scheme() ||
+      !url.SchemeIsHTTPOrHTTPS()) {
+    return;
+  }
+
+  if (session_) {
+    session_->Emit("preconnect", url.spec());
```

Comment:
> Also, should we pass along the allow_credentials and count values? I'm not sure what the client would do with them but there's presumably a reason they're here, right?

in chromium as I see, allow_credentials is always true when coming from loading predictor,
it can be false when coming from [message](https://cs.chromium.org/chromium/src/third_party/blink/renderer/core/html/parser/html_resource_preloader.cc?type=cs&sq=package:chromium&g=0&l=61) 

> The only place where the number of sockets is something other than 0 or 1 is in the loading predictor, which we don't seem to use here?

Right

Reply:
I made passing allow_credentials